design/healcheck-endpoint.md

```diff
@@ -0,0 +1,169 @@
+The implementation should include some experimentation, but the above 5 metrics should be enough to check the proper
+JCA state. The other metrics are available in [this document](https://docs.google.com/spreadsheets/d/1YGn4wwUupgQsFTmlm-5VXgLYkigTdy_TEnDhqFxNXXo/edit?usp=sharing).
+Another aspect is that JCA subsystem needs to actively probe the data source.
+This can be achieved by configuring `check-valid-connection-sql` and `background-validation`
+(see the [Wildfly Model manual](https://docs.wildfly.org/19/wildscribe/subsystem/datasources/data-source/ExampleDS/index.html)).
+
+#### Healthcheck API for Liveness and Readiness probes
+
+Since Undertow subsystem starts after the database migration, there's no way provide a working REST Health Endpoint
+during the database migration time. The long boot up time may also happen when loading up sessions from the database,
+or a remove Infinispan cluster. In both scenarios, it is necessary to implement a passing Liveness probe, and a failing
+Readiness probe. This will prevent Kubernetes from killing the Pods, but it will also unplug the pods from a Service
+and make them unavailable for the external consumption.
+
+One of the solutions is to use Wildfly's Liveness check available at `http://localhost:9990/health/live`. The check
+verifies if the server is up and if there are no boot errors (see [the code](https://github.com/wildfly/wildfly/blob/cf2ecf8cd575f425ba5cb254023fd6d44f2aa070/microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/HealthReporterService.java#L91-L93)).
+
+The Readiness probe will be invoked on Keycloak's REST interface - `https://<keycloak-url>:8443/auth/health`.
+The check will ensure that the REST interface is up and the database connectivity is fine.
+
+The REST response has been defined in [Microprofile Health Spec](https://download.eclipse.org/microprofile/microprofile-health-3.0-RC4/microprofile-health-spec.html#_appendix_a_rest_interfaces_specifications).
+There are 3 possible HTTP Status codes that might be returned from the endpoint:
+* 200 - Healthcheck passes
+* 500, 503 - Healthcheck fails
+
+The REST response body contains additional information helpful for debugging. A non-ready response call for
+`https://<keycloak-url>:8443/auth/health` might look like the following:
+
+```json
+{
+  "status": "DOWN",
+  "checks": [
+    {
+      "name": "org.keycloak.health.DatabaseCheck",
+      "status": "DOWN",
+      "data": {
+        "cause": "failing for 30s"
+      }
+    }
+  ]
+}
+```
+
+Having two Healthcheck endpoints is unfortunate. Once [WFLY-13931](https://issues.redhat.com/browse/WFLY-13931) gets
+implemented, it will be possible to migrate Keycloak's Readiness Check directly into Wildfly. For the time being,
+we a helper `readiness.sh` and `liveness.sh` scripts will be implemented in the Keycloak Container. This
+will enable us to transition between one and another solution transparently.
```

Comment:
I'm not really keen on this approach at all. Is there nothing better we can do?

Reply:
I hope we came to an agreement that the healthcheck complexity should go into the healthcheck code but the scripts are also requires as they enable the SRE Team (or any sort of Ops Team) to implement last-minute fixes and workarounds.

Let's keep the current approach - the Operator creates a ConfigMap with Bash scripts used for both probes. The Probes will invoke curl pointing to the Healthcheck REST. It's super flexible this way.